File: AlgorithmReview/Programmers/Lv2/캐시.swift
//
//  캐시.swift
//  AlgorithmReview
//
//  Created by 박소윤 on 2023/04/09.
//

import Foundation

func solution(_ cacheSize:Int, _ cities:[String]) -> Int {
    
    var ans = 0
    var cache = [String: Int]()
    
    if cacheSize == 0 {
        return cities.count * 5
    }
    
    for (i, c) in cities.enumerated() {
        let c = c.lowercased()
        if cache[c] != nil { //hit
            ans += 1
        } else {
            ans += 5
            if cache.count == cacheSize {
                let candidate = cache.min(by: { $0.value < $1.value })
                cache.removeValue(forKey: candidate!.key)
            }
        }
        cache[c] = i
    }
    return ans
}

File: Python/캐시.py
def solution(cacheSize, cities):
    
    answer = 0
    cache = {}
    
    if cacheSize == 0:
        return len(cities) * 5
    
    for i in range(0, len(cities)):
        data = cities[i].lower()
        if data in cache:
            answer += 1
        else:
            answer += 5
            if len(cache) == cacheSize:
                candidate = min(cache, key=cache.get)
                del cache[candidate]
                
        cache[data] = i
        
    return answer
